Make early stop the default for checkpoint and expose a way to disable

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #96923
* __->__ #96866
* #96862
* #96783
* #90105

Why did I choose context manager instead of per-call? Early stopping is not part of the model definition, and depending on how a particular model is used, e.g., with PT2 or not we may or may not want to disable early stopping.

soulitzer committed:
Make early stop the default for checkpoint and expose a way to disable

[ghstack-poisoned]

soulitzer committed:
Update on "Make early stop the default for checkpoint and expose a way to disable"




[ghstack-poisoned]

soulitzer committed:
Update on "Make early stop the default for checkpoint and expose a way to disable"


Why did I choose context manager instead of per-call? Early stopping is not part of the model definition, and depending on how a particular model is used, e.g., with PT2 or not we may or may not want to disable early stopping.

[ghstack-poisoned]

soulitzer committed:
Update on "Make early stop the default for checkpoint and expose a way to disable"


Why did I choose context manager instead of per-call? Early stopping is not part of the model definition, and depending on how a particular model is used, e.g., with PT2 or not we may or may not want to disable early stopping.

[ghstack-poisoned]

soulitzer committed:
Update on "Make early stop the default for checkpoint and expose a way to disable"


Why did I choose context manager instead of per-call? Early stopping is not part of the model definition, and depending on how a particular model is used, e.g., with PT2 or not we may or may not want to disable early stopping.

[ghstack-poisoned]

soulitzer committed:
Update on "Make early stop the default for checkpoint and expose a way to disable"


Why did I choose context manager instead of per-call? Early stopping is not part of the model definition, and depending on how a particular model is used, e.g., with PT2 or not we may or may not want to disable early stopping.

[ghstack-poisoned]

soulitzer committed:
Update on "Make early stop the default for checkpoint and expose a way to disable"


Why did I choose context manager instead of per-call? Early stopping is not part of the model definition, and depending on how a particular model is used, e.g., with PT2 or not we may or may not want to disable early stopping.

[ghstack-poisoned]

albanD commented on test/test_autograd.py:
```diff
@@ -10804,11 +10804,42 @@ def hook(*_unused_args):
+            return x.clone()
+
+        def fn(x):
+            # Since clone does not save anything, it is not recomputed iff
```
btw do we have tests to make sure the case where nothing needs to be recomputed works fine?
With non-reentrant it should be trivially working but still good to have a test.

albanD reviewed (approved):
Small questions on test but SGTM

soulitzer commented on test/test_autograd.py:
```diff
@@ -5624,18 +5624,18 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
             return verbose_mode, contextlib.nullcontext()
-        out = checkpoint(lambda x: x.sin(), x, use_reentrant=False, context_fn=context_fn)
-        self.assertEqual(verbose_mode.operators, ['sin.default'])
+        out = checkpoint(lambda x: x.exp(), x, use_reentrant=False, context_fn=context_fn)
```
Since this test is supposed to test usage of the `VerboseMode`, making it `exp` makes the recorded aten ops for the forward and recomputation look more similar. If it were still `sin`, since sin saves the input, recomputation is not needed at all, so nothing will show up in the trace.

(I suppose I could've disabled early stop as well.)

soulitzer committed:
Update on "Make early stop the default for checkpoint and expose a way to disable"


Why did I choose context manager instead of per-call? Early stopping is not part of the model definition, and depending on how a particular model is used, e.g., with PT2 or not we may or may not want to disable early stopping.

[ghstack-poisoned]
